src/main/java/com/socrata/datasync/validation/GISJobValidity.java

```diff
@@ -0,0 +1,637 @@
+            String[] supportedTypes;
+            if (component.equalsIgnoreCase("zip") || component.equalsIgnoreCase("latitude") || component.equalsIgnoreCase("longitude")) {
+                supportedTypes = new String[]{"text", "number"};
+            } else {
+                supportedTypes = new String[]{"text"};
+            }
+            String unsupportedType = getUnsupportedType(fieldname, supportedTypes, typeMapping);
+            if (unsupportedType != null) {
+                JobStatus status = JobStatus.PUBLISH_ERROR;
+                StringBuilder message = new StringBuilder("Unsupported Datatype: " + Utils.capitalizeFirstLetter(component) + " component  '" +
+                        locationName + "' is of type '" + unsupportedType + "'; " + component + " components require ");
+                if (supportedTypes.length == 1) message.append(" a ");
+                message.append("'" + supportedTypes[0] + "'");
+                for (int i=1; i<supportedTypes.length; i++)
+                    message.append(" or '" + supportedTypes[i] + "'");
+                message.append(" column");
+                if (supportedTypes.length > 1) message.append('s');
+                message.append(".\nPlease specify a column that matches a supported type.");
+                status.setMessage(message.toString());
+                return status;
+            }
+        }
+        return  JobStatus.VALID;
+    }
+
+    /**
+     * returns null if the component type is okay, otherwise the bad type.
+     */
+    private static String getUnsupportedType(String component, String[] supportedTypes, Map<String, String> typeMapping) {
+        String componentType = typeMapping.get(component);
+        if (component == null || componentType == null)
+            return null;
+
+        for (String okayType: supportedTypes) {
+            if (okayType.equalsIgnoreCase(componentType))
+                return null;
+        }
+        return componentType;
+    }
+
+    /**
+     * This check compares the information found in the headers to the columns found in the dataset.
+     * We check 3 things:
+     *   1) that the headers have a row identifier column if the dataset does
+     *   2) that the headers don't contain columns not in the dataset
+     *   3) that the headers aren't missing columns in the dataset IF there is no row identifier
+     * @param schema the soda-java Dataset that provides dataset info
+     * @param headers the columns that will be uploaded; taken from either the file or the control
+     *                file and minus the ignoredColumns
+     * @param csvFilename the file name, for printing purposes.
+     * @return
+     */
+    private static JobStatus checkColumnAgreement(FileTypeControl fileControl, PublishMethod method, Dataset schema, String[] headers, String csvFilename) {
+
+        String[] ignoredColumns = fileControl.ignoreColumns;
+
+        if (schema == null) return JobStatus.VALID;
```

Comment:
{}s, also consider adding a comment about why having no schema means it is valid.